Use Swift Equatable and Hashable conformances from ObjC

tbkka opened the pull request:
If a Swift type implements Equatable and/or Hashable and we then pass that object into ObjC, we want ObjC
`isEqual:` and `hashValue` to use that.  This allows ObjC code to put Swift objects into ObjC collections, including NSSet and NSDictionary.

* Support for Hashable struct/enum types was implemented in #4124
* Support for Equatable struct/enum types was implemented in #68720
* This PR implements support for Hashable and Equatable _class_ types

Caveats:
1. This does a lot of dynamic lookup work for each operation, so is inherently rather slow.  (Update:  Mike Ash suggested some changes that rely on heavily-cached results, so it's not as bad as I feared.)
2. This is a behavioral change to low-level support code.  There is a risk of breaking code that may be relying on the old behavior.
3. For a Swift type that is Equatable but not Hashable, the only sound way to forward `isEqual:` is to use a constant value for the hash.  This can cause major performance problems -- I've included a log message that warns whenever someone tries to hash such an object.

Resolves rdar://67093675

tbkka committed:
Use Swift Equatable and Hashable conformances from ObjC

If a Swift type implements Equatable and/or Hashable and
we then pass that object into ObjC, we want ObjC
`isEqual:` and `hashValue` to use that.  This allows
ObjC code to build ObjC collections of Swift objects.

* Support for Hashable struct/enum types was implemented in #4124
* Support for Equatable struct/enum types was implemented in #68720
* This implements support for Hashable and Equatable _class_ types

Caveats:
1. This does a lot of dynamic lookup work for each operation, so is
   inherently rather slow.  Unlike the struct/enum case, there is no convenient
   place to cache the conformance information, so it's not clear that there is a
   viable way to make it significantly faster.
2. This is a behavioral change to low-level support code.  There is a
   risk of breaking code that may be relying on the old behavior.

mikeash commented on stdlib/public/runtime/SwiftObject.mm:
```diff
@@ -373,11 +374,62 @@ + (BOOL)conformsToProtocol:(Protocol*)proto {
+    return YES;
+  }
+  // Both objects must be `SwiftObject` in ObjC
+  if (object_getClass(other) != object_getClass((id)self)) {
```
This is incorrect in the face of subclassing. If we get here, we have to go into the following code to search for a common parent. However, the classes will be the same in the common case, so short-circuiting the common-parent code when the classes are equal would be productive.

tbkka replied:
The intent here is to make sure that `other` really came from Swift.

My understanding is that there are three different types for each object here:
* The type that's in the Obj-C header in memory -- I thought this was always _SwiftObject
* The Obj-C type returned from `-class` -- this is dynamically looked up in _SwiftObject's -class implementation
* The Swift type of the object

I thought that `object_getClass()` always returned the first of these.

mikeash replied:
Ah, not quite. The root class is `SwiftObject`, but the instance's class is the Swift class. Swift classes are ObjC classes. So if you have:

```
class C {}
object_getClass(C())
```

This will return `C`.

There's one case where things differ a little bit, which is pure ObjC classes. This is because pure ObjC classes are not valid Swift metadata, so they get a wrapper on the Swift end of things:

```
object_getClass(NSObject()) // returns the NSObject class
type(of: NSObject()) // returns a Swift wrapper metadata that points to NSObject
```

But that's the only case where things are like that.

tbkka replied:
Thanks for that clarification.  I believe I have that sorted out now.

mikeash commented on test/stdlib/SwiftObjectNSObject.swift:
```diff
@@ -39,8 +39,24 @@ class D : C {
   @objc override class func cClassOverride() -> Int { return 8 }
 }
 
+class E : Equatable {
```
Would be good to have two subclasses, and test all combinations of: two instances of same subclass, two instances of different subclasses, instance of subclass and instance of superclass.

tbkka replied:
I've significantly extended the test cases.  This mostly involved adding some helpers that checked the Swift behavior then delegated to an Obj-C test function that verified the behavior matches Swift.

lorentey commented on stdlib/public/runtime/SwiftObject.mm:
```diff
@@ -373,11 +374,62 @@ + (BOOL)conformsToProtocol:(Protocol*)proto {
+    return NO;
+  }
+
+  return _swift_stdlib_Equatable_isEqual_indirect(
```
This will need to be more careful, unfortunately. NSObject subclasses must always implement hashing, and `-[NSObject hash]` must always be consistent with `-[NSObject isEqual:]`. 

If the Swift type only implements `Equatable`, then `-hash` falls back to using the object's identity above, but in this implementation `isEqual:` would still forward to `T.==`. This would allow `[a isEqual: b] && a.hash != b.hash`, which would violate the NSObject protocol and wreak havoc.

So `-[SwiftObject isEqual:]` must emulate `hash`, and only forward to `==` iff the underlying type is `Hashable`; otherwise `isEqual:` must continue to compare object identities, whether or not the type is `Equatable`.

tbkka replied:
Alternatively, could hashing default to always return 0?   That would also continue to meet the hash/equality consistency requirement.

tbkka replied:
Apparently, NSObject's default implementation of `hashValue` simply returns `self`.  So it would make sense for that to continue to be the default for `SwiftObject` as well, in which case Karoy's right that equality can only be driven from the Hashable conformance.

mikeash replied:
It’s a little weird since ObjC doesn’t enforce this requirement except in the sense that collections behave strangely if you mess it up. In practice, people will break this requirement knowing that they don’t use their objects in hashing collections. (And then they’ll forget about it and use a hashing collection and be sad.)

More strongly enforcing this requirement would fit with Swift’s greater emphasis on safety.

We could have SwiftObject return a 0 hash for the specific case of a type that is Equatable but not Hashable….

lorentey replied:
Returning 0 from `-hash` would technically work! (It's not generally a good idea, though. 😅)

tbkka replied:
Let me try that approach.   The alternative is to ignore the user-defined `==` which isn't generally a good idea either.  😄

tbkka replied:
I've changed `-hash` to return the constant 1.

(Rationale:  people sometimes mix hashvalues of their properties via addition, and `0` has the drawback of not distinguishing types with different numbers of properties in such cases.  I don't see an argument for or against any other value, so arbitrarily chose `1`.)

lorentey commented on test/stdlib/Inputs/SwiftObjectNSObject/SwiftObjectNSObject.m:
```diff
@@ -159,12 +162,24 @@ void TestSwiftObjectNSObject(id c, id d)
   expectFalse([S_meta isEqual:C_meta]);
 
+  // Check that ObjC isEqual delegates to Equatable
+  expectTrue([e1a isEqual: e1b]);
+  expectFalse([e1a isEqual: e2]);
+  expectFalse([e1b isEqual: e2]);
```
```suggestion
  expectTrue([e1a isEqual: e1a]);
  expectFalse([e1a isEqual: e1b]);
  expectFalse([e1a isEqual: e2]);
  expectFalse([e1b isEqual: e2]);
```

tbkka replied:
See above.  I now use a constant for the hashValue of non-Hashable types, which I believe means that we can continue to bridge Swift `==` in these cases.

tbkka committed:
Fix a number of issues (review feedback)

* Fix how we were looking for common parent class
* Use a constant hashValue for types that are Equatable but not Hashable

tbkka commented:
@swift-ci Please test

mikeash commented on stdlib/public/runtime/SwiftObject.mm:
```diff
@@ -390,36 +395,39 @@ - (BOOL)isEqual:(id)other {
-    if (selfParents.find(m) != selfParents.end()) {
-      parentMetadata = m;
-      break;
+  if (selfMetadata == otherMetadata) {
```
In theory, it should be good enough to do `swift_conformsToProtocol` on `selfMetadata`, use `findConformingSuperclass` to find the type that actually contains the conformance, then see if `otherMetadata` is that type or a subtype. We don't allow subclasses to provide a different conformance than a superclass, so this would find the only allowable common conformance, if there is one. It would also avoid traversing the entire superclass chain when there is a conformance, and would cache the negative result to avoid traversing the superclass chain more than once when there isn't.

However, it's still possible to have multiple conformances in a hierarchy, even though we don't allow it, so there are situations where that approach would provide a different answer. So I'm not sure if it's actually a good idea. Just putting it out there as food for thought.

tbkka replied:
I tried your idea.  Let me know if this looks right to you.  In particular, thanks for pointing out that some of these lookups get cached.

mikeash replied:
Looks good! I should also note that we do have a warning when we find multiple conformances in a hierarchy, so there's at least some feedback when that goes wrong: https://github.com/apple/swift/blob/main/stdlib/public/runtime/ProtocolConformance.cpp#L1119

tbkka committed:
Allow "incomplete hashing" when testing ObjC bridged objects

Swift objects can implement Equatable without implementing
Hashable.  Obj-C NSObject always provides both.

Now that we bridge Swift Equatable to Obj-C `-isEqual:`,
we have to be conservative with how we expose `-hash`
for Swift types that are Equatable but not Hashable.
This looks like "incomplete hashing" of such types.

tbkka committed:
Use a simpler strategy for finding the Equatable conformance to use

tbkka committed:
Handle ObjC `-hash` differently for Equatable and non-Equatable types

For an Equatable type, we need a hash implementation that
is compatible with any possible definition of `==`.
Conservatively, that means `-hash` must return a constant.

For non-Equatable types, we know that `==` is identity based,
so we can get better hash behavior by using the object address.

Caveat:  This means that non-Equatable types will do two
protocol conformance checks on every call to `hash`.

tbkka commented on stdlib/public/runtime/SwiftObject.mm:
```diff
@@ -373,11 +374,72 @@ + (BOOL)conformsToProtocol:(Protocol*)proto {
+  }
+
+  // Legacy default for types that are neither Hashable nor Equatable.
   return (NSUInteger)self;
```
@lorentey @mikeash I'm a little torn on this one.  I experimentally changed `-hash` to have three cases:
* Use the Swift Hashable conformance if present
* Return a constant if this is Swift Equatable but not Swift Hashable (this is necessary to be compatible with an arbitrary `==`)
* Return `self` if this is neither Swift Equatable nor Swift Hashable.

These are arguably the optimal hash values for each case, but there's a lot of dynamic overhead here.  Should I just go back to returning a constant for both of the last two cases?

mikeash replied:
The overhead shouldn't be too bad. Looking up a cached conformance isn't as fast as, say, an `objc_msgSend`, but it's pretty quick. If this is somehow a performance issue for someone, they can pretty easily resolve it by implementing their own `hash` and `isEqual:`.

tbkka replied:
So it sounds like the three-way split here seems the best option in terms of semantics.   I agree that a constant hash is a bad idea, but I don't see any other really viable option for this specific case.  I could however add a log message here calling out that someone tried to hash a Swift type that isn't Hashable and that this can lead to performance problems.  (Trapping is tempting, but that feels like a Swift answer, not an Obj-C answer.  🤔)

> ... we'll want __SwiftValue to implement exactly the same logic ...
Of course.  I have those changes locally and will submit them as a separate PR once we've nailed down the behavior we want here.

lorentey replied:
Logging a runtime message is a really good idea! Do we have a rate-limited logging facility?

(We'd want to use it from Swift parts of the stdlib too.)

tbkka replied:
Added.

mikeash replied:
I don't think we have one already. Rather than rate-limiting, it might be good to limit it to once per class. We'd have to keep a set of classes we've logged for, but if the whole point is to call out performance problems, maybe that's OK.

tbkka replied:
Any good examples of handling such a set (thread-safety, etc)?

mikeash replied:
Unchecked continuation validation is pretty much this. Only difference is that it also removes items, but you don't need that part. Link: https://github.com/apple/swift/blob/c1e892418fff4a91d9abf5e957b21f748fd28e7a/stdlib/public/Concurrency/Task.cpp#L1280

tbkka replied:
I went ahead with a std::unordered_set for this, copying the example from continuation validation.  I also updated the test to verify that the runtime does in fact emit one message for each type that hits this particular bit of code.

mikeash replied:
Seems fine to use `unordered_set`. I don't have a good sense of the tradeoffs between the two but `unordered_set` clearly does work. We can always switch it if trouble arises.

lorentey commented:
Note: It really would be rather risky to land this change, as it is readily observable, and it can change a program's semantics in ways that are subtle enough not to be immediately obvious.

The correct workaround for clients with Equatable types that need `-isEqual:` to forward to `==` is to also conform their type to `Hashable`, as NSObject requires hashing. That is usually straightforward to implement, as long as the root cause is clear. (Granted, it normally isn't. 😞)

Instead of returning a bogus constant, we could instead choose to trap / throw an exception if someone tries to call `-hash` on an Equatable-but-not-Hashable type. That would make the hashing problem (and its solution) immediately obvious, but it would be quite disruptive at this point, even with a linked-on-or-after check. It would also only deal with the `-hash` problem: it wouldn't help with the semantics of `isEqual:`. (If we were designing this today, perhaps something like this would've been a better option, though.)

tbkka committed:
Log a warning if Obj-C `-hash` is called on a Swift type that is Equatable but not Hashable

The only reasonable way to handle this can lead to severe
performance drop-offs.  We should make sure the
user is aware of this.

tbkka commented on stdlib/public/runtime/SwiftObject.mm:
```diff
@@ -373,11 +374,77 @@ + (BOOL)conformsToProtocol:(Protocol*)proto {
+    warning(0,
+	    "Obj-C `-hash` invoked on a Swift object of type %s that is not Hashable; "
+	    "this can lead to severe performance problems",
+	    clsName);
```
It would be nice to be able to rate-limit this.   @mikeash Any thoughts?

tbkka committed:
Log a message when Obj-C `-hash` is invoked for a Swift object that is Equatable but not Hashable

This is worth telling people about:

Since we're being asked for the hash value, that probably means someone is
trying to put this object into some form of set or dictionary.  But we don't know
anything about the Equatable implementation, so the only valid hash value we can
possibly return is a constant.  And hash table implementations typically degrade
into unsorted lists when provided constant hash values.

Note: In order to avoid hammering the logs, we only emit the
warning once for each class that hits this path.

tbkka committed:
Ooops! Missing includes

tbkka commented:
I believe I've now addressed all the concerns raised by @mikeash and @lorentey.   There is admittedly some risk here due to the change in behavior, but I feel like I've mitigated that about as much as I can.  Any further concerns?

tbkka commented:
@swift-ci Please test

tbkka commented:
@lorentey I'd prefer not to merge this if you still have concerns.  (And your last "official" review did request changes.)

tbkka commented:
Merged #69618 to use the three-way approach for SwiftValue `-hash` implementation.
